### ISSUE TITLE
Implementation details in CH21

### ISSUE THREAD

**Issue by siuszy** [fluentpython/example-code-2e#16]
I deployed the local server according to SETTING UP TEST SERVERS, finding two small suspected errors both in **slow_server.py**.

The first one is  about the **parser** in the main fucncion(**Line 59**). The sokcet module requires the parameter bind to be _str, bytes, or bytearray_ , so I add a default value as the other part. (This problem is maybe linked to the python environment thing)

The second one is in the **doGet** function of **SlowHTTPRequestHandler**(**Line41**). I get an AttributeError when I first ran the ERROR option, which told me that the HTTPStatus did not have this attribute **IM_A_TEAPOT**, so maybe we could just throw a number(418 according to the book) instead?

**Comment by leorochael** [comment]
Hi @siuszy, The `bind` cmdline argument in `slow_server.py` is passed into `http.server.test()`, from there into `http.server._get_best_family`, and finally into `socket.getaddrinfo()`.

[The documentation](https://docs.python.org/3/library/socket.html#socket.getaddrinfo) specifies that the `host` parameter from `getaddrinfo` accepts `str`, `bytes` or `None`.

And `None` is exactly what the value of `args.bind` will be if the `--bind` parameter is not specified and its definition has no `default` value, according to the definition of the `parser.add_argument()` found in the line 59 you pointed out.

The example runs fine when `args.bind` is None, outputing:

    Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...

Which is exactly what is intended.

Can you please specify in detail what problem do you see exactly when the `--bind` argument has no `default` and is no passed in your environment?

**Comment by siuszy** [comment]
The document said that None is accepted, but the error message thrown by the **socket.bind()** showed that NoneType is not allowed. I thought maybe when the default value is None, the undergo procedure is somewhat blocked by my environment(Anaconda)?

**Comment by ramalho** [comment]
@siuszy we are unable to reproduce the problem you are having.

1. what is the Python version you are using (use `python -V` on the command line).
2. please reproduce precisely the complete traceback that appears in your terminal, starting with the complete command line that you are using to run the `slow_server.py` script, and ending with the line with the final exception message.

Thank you.

**Comment by leorochael** [comment]
I managed to track down the problem to Python version 3.7:

```
$ python slow_server.py 
Traceback (most recent call last):
  File "slow_server.py", line 92, in <module>
    bind=args.bind,
  File "/usr/lib/python3.7/http/server.py", line 1230, in test
    with ServerClass(server_address, HandlerClass) as httpd:
  File "/usr/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "slow_server.py", line 85, in server_bind
    return super().server_bind()
  File "/usr/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
TypeError: str, bytes or bytearray expected, not NoneType
```

The default value for the `bind=` parameter of `http.server.test` in that version is the empty string instead of `None`.

**Comment by leorochael** [comment]
Adding a `default='',` to the argument parser configuration for the `bind` parameter fixes it for Python 3.7:

```python
    parser.add_argument('--bind', '-b', metavar='ADDRESS', default='',
                        help='Specify alternate bind address '
                             '[default: all interfaces]')
```

But it breaks for more recent versions, which expects `None` as the default parameter, so it's not an acceptable solution.

A more complicated solution for stradling these versions would be to try to fetch the default `bind=` value from the `http.server.test()` function, or to try not to pass the `bind=` argument if the value is not given on the command line.

But I believe both these alternatives are out of scope for the code in the book, which should strive for simplicity.

**Comment by siuszy** [comment]
@ramalho @leorochael sorry for causing this trouble and haven't clarified the trouble.!  Last comment showed the exact trouble I met and further gave me the answer, thank you so much! I tried to locate the exact function that caused the trouble but I was totally lost (thought it was something in the socket module).
